proposals/interface-types/CanonicalABI.md

```diff
@@ -0,0 +1,1203 @@
+
+          # Recursively read the payload, if present for this case. Note that
+          # we align the pointer to the whole variant's alignment to ensure that
+          # all payloads start at the same address.
+          if payload:
+            offset = align_to(offset, align(direction, variant $cases))
+            yield from lift_from_memory(direction, dst, ptr + offset, payload)
+
+          yield(variant_end<$cases>)
+        }
+      }
+    }
+
+    # defer validation and such to `lift`
+    handle<$resource> => {
+      val = read_4_bytes(src, ptr)
+      yield from lift(direction, src, handle<$resource>, [val as wasm_i32])
+    }
+
+    push-buffer<$ty> | pull-buffer<$ty>(val) => {
+      assert(direction == "import") # see comments above
+      handle = read_4_bytes(src, ptr)
+      offset = read_4_bytes(src, ptr + 4)
+      len = read_4_bytes(src, ptr + 8)
+      yield from lift(direction, src, ty, [handle as wasm_i32, offset as wasm_i32, len as wasm_i32])
+    }
+```
+
+## Lowering
+
+The `lower` function here takes three parameters:
+
+* `direction` - one of `"export"` or `"import"` which indicates for which
+  purposes the lowering is happening: lowering an export means lowering
+  a parameter, lowering an import means lowering a result.
+* `dst` - this is an instance where the value is being lowered into. This could
+  be a WebAssembly instance or the host itself.
+* `gen` - a generator representing the interface types values that's being
+  lowered. This is expected to be the generator returned from calling `lift`
+  above.
+
+This defines how an interface types value is encoded into linear memory and
+WebAssembly values.
+
+```
+def lower(direction, dst, gen):
+  match gen.next()
+    f32(val) => [val as wasm_f32]
+    f64(val) => [val as wasm_f64]
+
+    # sign-extended
+    s8(val) => [val as wasm_i32]
+    s16(val) => [val as wasm_i32]
+
+    # zero-extended
+    u8(val) => [val as wasm_i32]
+    u16(val) => [val as wasm_i32]
+
+    s32(val) => [val as wasm_i32]
+    u32(val) => [val as wasm_i32]
+    s64(val) => [val as wasm_i64]
+    u64(val) => [val as wasm_i64]
+
+    char(val) => [val as wasm_i32]
+
+    # Note that `val` here is guaranteed to be a valid list of unicode scalar
+    # values so encoding here should always succeed.
+    string(val) => {
+      base = dst.canonical_abi_realloc(NULL, 0, 1, utf8_byte_length(val))
+      assert_or_trap(base + utf8_byte_length(val) <= dst.memory.len)
+      encode_utf8_string(dst, base, val)
+      [base as wasm_i32, val.utf8_len() as wasm_i32]
+    }
+
+    # List are represented by their pointer/length as 32-bit integers. Note that
+    # the `canonical_abi_realloc` can be elided if `dst` is the host itself since
+    # the host can retain the raw pointers into the wasm module's memory if it
+    # likes.
+    list_start<$ty>(len) => {
+      base = dst.canonical_abi_realloc(NULL, 0, align(direction, $ty), len * size(direction, $ty))
```

Comment:
Is this following C `realloc` behavior of using a `NULL` original pointer to make `realloc` behave like `malloc`? If so, that's a little tricky for wasm, where the null page isn't guarded, and programs (not written in Rust/C/etc.) could potentially allocate data there. Would it make sense to define a `canonical_abi_malloc` for this case?

Reply:
Can't we just have the the canonical API define zero as the null pointer?   It doesn't dictate an entire null page, just the specific value zero... So it would add single byte of overhead per module to be compatible with such an ABI wouldn't it?

Reply:
That all seems fine to me, but I'm not sure how to best fit it in. Are you thinking that this would clarify what a zero-sized allocation would look like? Or other words somewhere else perhaps?

Reply:
I think you could describe it in the Memory Intrinsics section, along with the description of `canonical_abi_realloc`. Something along the lines of "Unlike C's `realloc`, zero-length allocations need not have unique addresses" might be enough for now, and we can pin down the specific requirements later.